### PR TITLE
add revision and checkPut

### PR DESCRIPTION
**Pull request by ashley-taylor**
revision is added to the table object.
Each put or link operation will increment this value by one

added a checkPut method that only passes if the database has the same revision number as the version specified.

Would you also like a
checkDelete
checkLink / checkLinks methods?

### CODE REVIEW

File: graphql-database-manager-dynamo/src/main/java/com/fleetpin/graphql/database/manager/dynamo/DynamoDb.java [hunk 1]
```diff
@@ -28,15 +26,15 @@
...
+
 import software.amazon.awssdk.services.dynamodb.DynamoDbAsyncClient;
 import software.amazon.awssdk.services.dynamodb.model.*;
-import software.amazon.awssdk.utils.CompletableFutureUtils;
 
 public final class DynamoDb extends DatabaseDriver {
-    private final AttributeValue GLOBAL = AttributeValue.builder().s("global").build();
+    private static final AttributeValue REVISION_INCREMENT = AttributeValue.builder().n("1").build();
+
+	private final AttributeValue GLOBAL = AttributeValue.builder().s("global").build();
```
Comment (Aliics) [inline graphql-database-manager-dynamo/src/main/java/com/fleetpin/graphql/database/manager/dynamo/DynamoDb.java:37]: Just a nitpick, but adjust the indentation.

File: graphql-database-manager-dynamo/src/main/java/com/fleetpin/graphql/database/manager/dynamo/DynamoDb.java [hunk 2]
```diff
@@ -130,11 +131,37 @@ public DynamoDb(ObjectMapper mapper, List<String> entityTables, DynamoDbAsyncCli
...
+        			mutator.conditionExpression("attribute_not_exists(id)");
+        		}else {
+        			Map<String, AttributeValue> variables = new HashMap<>();
+        			variables.put(":revision", AttributeValue.builder().n(Integer.toString(revision)).build());
+					//check exists and matches revision
+        			mutator.expressionAttributeValues(variables);
+        			mutator.conditionExpression("revision = :revision");
+        		}
+        		
+        	}
```
Comment (Aliics) [inline graphql-database-manager-dynamo/src/main/java/com/fleetpin/graphql/database/manager/dynamo/DynamoDb.java:155]: Indentation fix here too pls. :)

File: graphql-database-manager-test/src/test/java/com/fleetpin/graphql/database/manager/test/DynamoDbRevisionTest.java [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+public class DynamoDbRevisionTest {
+	
+	@TestDatabase
+	public void testCreateNewObject(final Database db) throws InterruptedException, ExecutionException {
+		var entry = new SimpleTable("1", "garry");
+		db.checkPut(entry).get();
+		var entries = db.query(SimpleTable.class).get();
+		Assertions.assertEquals(1, entries.size());
+		Assertions.assertEquals("garry", entries.get(0).name);
+
```
Comment (Aliics) [inline graphql-database-manager-test/src/test/java/com/fleetpin/graphql/database/manager/test/DynamoDbRevisionTest.java:35]: Could we assert against the revision here?

File: graphql-database-manager-test/src/test/java/com/fleetpin/graphql/database/manager/test/DynamoDbRevisionTest.java [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+		var cause = Assertions.assertThrows(ExecutionException.class, () -> db.checkPut(entry).get());
+		Assertions.assertEquals(RevisionMismatchException.class, cause.getCause().getClass());
+	}
+
+	@TestDatabase
+	public void testIncrementsAfterExists(final Database db) throws InterruptedException, ExecutionException {
+		var entry = new SimpleTable("1", "garry");
+		
+		entry = db.checkPut(entry).get();
+		entry = db.checkPut(entry).get();
```
Comment (Aliics) [inline graphql-database-manager-test/src/test/java/com/fleetpin/graphql/database/manager/test/DynamoDbRevisionTest.java:56]: Maybe assert the revision after each put? It at least helps me follow what sort of state the row is in.

File: graphql-database-manager-test/src/test/java/com/fleetpin/graphql/database/manager/test/DynamoDbRevisionTest.java [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+			var cause = Assertions.assertThrows(ExecutionException.class, () -> db.checkPut(e).get());
+			Assertions.assertEquals(RevisionMismatchException.class, cause.getCause().getClass());
+		}
+		
+		{
+			entry.setRevision(0);
+			var e = entry;
+			var cause = Assertions.assertThrows(ExecutionException.class, () -> db.checkPut(e).get());
+			Assertions.assertEquals(RevisionMismatchException.class, cause.getCause().getClass());
+		}
```
Comment (Aliics) [inline graphql-database-manager-test/src/test/java/com/fleetpin/graphql/database/manager/test/DynamoDbRevisionTest.java:85]: These really show that it's working nicely. :)

File: graphql-database-manager-dynamo/src/main/java/com/fleetpin/graphql/database/manager/dynamo/DynamoDb.java [hunk 6]
```diff
@@ -405,37 +436,43 @@ public int maxBatchSize() {
...
                     }
-                    return CompletableFuture.completedFuture(null);
+                    return CompletableFuture.completedFuture(r);
                 }).thenCompose(a -> a);
         future = future.thenCombine(destination, (a, b) -> b);
         return future.thenApply(response -> {
+        	entity.setRevision(Integer.parseInt(response.attributes().get("revision").n()));
             return entity;
         });
     }
```
Comment (Aliics) [inline graphql-database-manager-dynamo/src/main/java/com/fleetpin/graphql/database/manager/dynamo/DynamoDb.java:478]: This method is now a behemoth in size...
I'm finding it a little hard to follow the whole thing.
Could we cut it up into smaller pieces with some aptly named methods?

Comment (Aliics) [reply graphql-database-manager-dynamo/src/main/java/com/fleetpin/graphql/database/manager/dynamo/DynamoDb.java:478]: It only selected one line, but it's for the whole "link" method.

File: graphql-database-manager-dynamo/src/main/java/com/fleetpin/graphql/database/manager/dynamo/DynamoDb.java (hunk 2 above)
Comment (ashley-taylor) [reply graphql-database-manager-dynamo/src/main/java/com/fleetpin/graphql/database/manager/dynamo/DynamoDb.java]: fixed. Had tabs not spaces :(

File: graphql-database-manager-dynamo/src/main/java/com/fleetpin/graphql/database/manager/dynamo/DynamoDb.java (hunk 1 above)
Comment (ashley-taylor) [reply graphql-database-manager-dynamo/src/main/java/com/fleetpin/graphql/database/manager/dynamo/DynamoDb.java]: fixed. Had tabs not spaces :(

File: graphql-database-manager-core/src/main/java/com/fleetpin/graphql/database/manager/Table.java [hunk 7]
```diff
@@ -24,6 +24,7 @@
 public abstract class Table {
 
 	private String id;
+	private int revision;
```
Comment (joshua-rule) [inline graphql-database-manager-core/src/main/java/com/fleetpin/graphql/database/manager/Table.java:27]: Should we use a long for the revision?

File: graphql-database-manager-dynamo/src/main/java/com/fleetpin/graphql/database/manager/dynamo/DynamoDb.java [hunk 8]
```diff
@@ -99,14 +96,17 @@ public DynamoDb(ObjectMapper mapper, List<String> entityTables, DynamoDbAsyncCli
         if (entity.getCreatedAt() == null) {
             setCreatedAt(entity, Instant.now()); //if missing for what ever reason
         }
+        final int revision = entity.getRevision();
```
Comment (joshua-rule) [inline graphql-database-manager-dynamo/src/main/java/com/fleetpin/graphql/database/manager/dynamo/DynamoDb.java:99]: We could have the table class have an incrementRevision method.

Comment (ashley-taylor) [reply graphql-database-manager-dynamo/src/main/java/com/fleetpin/graphql/database/manager/dynamo/DynamoDb.java:99]: I can't hide it from the consumers. Feel should only be visible to the DB layer.

We do need to expose setRevision for usecase when you contain the revision in your mutation to confirm editing latest.

**Comment by justin-mccormack** [comment]
Did you consider overloading existing methods?
e.g. `public <T extends Table> CompletableFuture<T> put(T entity, Boolean checkRevision = false)`

Maybe consider using a property name that isn't likely to clash with application property name.
For example, MongoDB uses `_v`.

Could also be useful to enable/force checking by default -- i.e. revision checking for all `put` operations on the `Location` table to avoid multiple users making clashing edits). Could this be specified as an option on the table itself?

File: graphql-database-manager-test/src/test/java/com/fleetpin/graphql/database/manager/test/DynamoDbRevisionTest.java (hunk 3 above)
Comment (ashley-taylor) [reply graphql-database-manager-test/src/test/java/com/fleetpin/graphql/database/manager/test/DynamoDbRevisionTest.java:36]: done

File: graphql-database-manager-core/src/main/java/com/fleetpin/graphql/database/manager/Table.java (hunk 7 above)
Comment (ashley-taylor) [reply graphql-database-manager-core/src/main/java/com/fleetpin/graphql/database/manager/Table.java]: done.

File: graphql-database-manager-dynamo/src/main/java/com/fleetpin/graphql/database/manager/dynamo/DynamoDb.java (hunk 6 above)
Comment (ashley-taylor) [reply graphql-database-manager-dynamo/src/main/java/com/fleetpin/graphql/database/manager/dynamo/DynamoDb.java:470]: done

File: graphql-database-manager-test/src/test/java/com/fleetpin/graphql/database/manager/test/DynamoDbRevisionTest.java (hunk 4 above)
Comment (ashley-taylor) [reply graphql-database-manager-test/src/test/java/com/fleetpin/graphql/database/manager/test/DynamoDbRevisionTest.java:58]: done
